payments/providers/cpu_ceepos.py

```diff
@@ -0,0 +1,361 @@
+            }
+        )
+
+    def payload_add_checksum(self, payload):
+        """
+        Attaches the checksum to the payload.
+
+        The checksum is calculated from a string that consists of the values of
+        the parameters in the payload and the source system specific secret key.
+        """
+        checksum_calc_values = []
+        for key, value in payload.items():
+            if isinstance(value, list):
+                for item in value:
+                    checksum_calc_values.extend(item.values())
+            else:
+                checksum_calc_values.append(value)
+        payload["Hash"] = self.calculate_checksum(checksum_calc_values)
+
+    def validate_ceepos_message_payload(self, request=None, response=None):
+        """
+        Validates that message payload received from CeePos has not been tampered with.
+        The received message is either a GET or POST request (success/notify) or
+        a response to the initiate payment request.
+
+        If the checksum calculated is the same as the Hash parameter in the request/response,
+        the message has been received intact and directly from CeePos.
+        """
+
+        REQUEST_CHECKSUM_PARAMS = (
+            "Id",
+            "Status",
+            "Reference",
+            "PaymentMethod",
+            "PaymentSum",
+            "Timestamp",
+            "PaymentDescription",
+        )
+        RESPONSE_CHECKSUM_PARAMS = (
+            "Id",
+            "Status",
+            "Reference",
+            "Action",
+            "PaymentAddress",
+            "PaymentExpires",
+        )
+
+        is_valid = True
+
+        if request:
+            if request.method == "GET":
+                # success request
+                checksum_received = request.GET["Hash"]
+                checksum_calc_values = [
+                    request.GET[param_name]
+                    for param_name in REQUEST_CHECKSUM_PARAMS
+                    if param_name in request.GET
+                ]
+            elif request.method == "POST":
+                # notify request
+                checksum_received = request.POST["Hash"]
+                checksum_calc_values = [
+                    request.POST[param_name]
+                    for param_name in REQUEST_CHECKSUM_PARAMS
+                    if param_name in request.POST
+                ]
+        elif response:
+            # initiate payment response
+            checksum_received = response["Hash"]
+            checksum_calc_values = [
+                response[param_name]
+                for param_name in RESPONSE_CHECKSUM_PARAMS
+                if param_name in response
+            ]
+
+        correct_checksum = self.calculate_checksum(checksum_calc_values)
+
+        if not hmac.compare_digest(checksum_received, correct_checksum):
+            LOG.warning('Incorrect checksum "{}".'.format(checksum_received))
+            is_valid = False
+
+        return is_valid
```

Comment:
Why the variable? Now I have to spend more time to track if the is_valid is set to False somewhere else too. It would nicer to just `return True` here and `return False`  on the line above.

Reply:
That is true, I'll adjust this

Reply:
Changed this.